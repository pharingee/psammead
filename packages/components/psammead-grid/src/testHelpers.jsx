import React from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import Paragraph from '@bbc/psammead-paragraph';
import { cyrillicAndLatin } from '@bbc/gel-foundations/scripts';
import Image from '@bbc/psammead-image';
import { GEL_SPACING } from '@bbc/gel-foundations/dist/spacings';

export const ExampleParagraph = ({ identifier }) => (
  <Paragraph script={cyrillicAndLatin} service="news">
    {identifier}This is a long paragraph that will wrap for several lines. This
    is a long paragraph that will wrap for several lines. This is a long
    paragraph that will wrap for several lines. This is a long paragraph that
    will wrap for several lines. This is a long paragraph that will wrap for
    several lines. This is a long paragraph that will wrap for several lines.
    This is a long paragraph that will wrap for several lines. This is a long
    paragraph that will wrap for several lines.
  </Paragraph>
);

export const ExampleFigure = styled.figure`
  margin: 0;
  padding: 0;
`;

const ImageSpacing = styled.div`
  margin: 0;
  padding: 0 0 ${GEL_SPACING} 0;
`;

export const ExampleImage = () => (
  <ImageSpacing>
    <Image
      alt="Robert Downey Junior in Iron Man"
      src="https://ichef.bbci.co.uk/news/660/cpsprodpb/11897/production/_106613817_999_al_.jpg"
      width="640"
    />
  </ImageSpacing>
);

ExampleParagraph.propTypes = {
  identifier: string,
};
ExampleParagraph.defaultProps = {
  identifier: '1',
};
