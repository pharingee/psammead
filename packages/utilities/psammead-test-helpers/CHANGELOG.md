# Psammead Test Helpers Changelog

<!-- prettier-ignore -->
| Version | Description |
|---------|-------------|
| 3.0.2 | [PR#2166](https://github.com/bbc/psammead/pull/2166) remove async keyword from test helper to fix regenerator runtime plugin error |
| 3.0.1 | [PR#2123](https://github.com/bbc/psammead/pull/2125) Removes use of async/await |
| 3.0.0 | [PR#2104](https://github.com/bbc/psammead/pull/2104) add function to detect and render helmet for use in snapshots |
| 2.0.2 | [PR#1994](https://github.com/bbc/psammead/pull/1994) Add comment to clarify new shouldMatchSnapshot logic |
| 2.0.1 | [PR#1958](https://github.com/bbc/psammead/pull/1958) Update `shouldMatchSnapshot` to add firstChild logic |
| 2.0.0 | [PR#1917](https://github.com/bbc/psammead/pull/1917) Remove react-test-renderer and shallow rendering |
| 1.3.2 | [PR#1919](https://github.com/bbc/psammead/pull/1919) Accidental version bump |
| 1.3.1 | [PR#1849](https://github.com/bbc/psammead/pull/1849) Update README.md |
| 1.3.0 | [PR#1846](https://github.com/bbc/psammead/pull/1846/) Import test helper functions from simorgh |
| 1.2.1 | [PR#1803](https://github.com/bbc/psammead/pull/1803/) Patches broken links on badges in documentation |
| 1.2.0 | [PR#1794](https://github.com/bbc/psammead/pull/1794) Add david dependency badges |
| 1.1.1 | [PR#1682](https://github.com/bbc/psammead/pull/1682) Move all dev dependencies to top level package.json |
| 1.1.0   | [PR#1233](https://github.com/bbc/psammead/pull/1233) Add ESM modules entry |
| 1.0.2   | [PR#1083](https://github.com/bbc/psammead/pull/1083) Use `jest-styled-components@6.3.3` |
| 1.0.1   | [PR#732](https://github.com/bbc/psammead/pull/732) Use @testing-library/react for generating snapshots |
| 1.0.0   | [PR#679](https://github.com/bbc/psammead/pull/679) Bump version number |
| 0.3.4   | [PR#424](https://github.com/bbc/psammead/pull/424) Add Snyk badge to readme |
| 0.3.3   | [PR#323](https://github.com/BBC/psammead/pull/323) Update readme storybook badge |
| 0.3.2   | [PR#264](https://github.com/BBC/psammead/pull/319) Resolving package-lock issues. |
| 0.3.1   | [PR#264](https://github.com/BBC/psammead/pull/264) Resolving package-lock issues. |
| 0.3.0   | [PR#224](https://github.com/BBC-News/psammead/pull/224) Adds a helper method for testing package export values and uses it to test its own exports. |
| 0.2.1   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
