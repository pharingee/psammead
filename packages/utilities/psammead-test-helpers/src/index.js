import { render } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';
import renderWithHelmet from './renderWithHelmet';

export const shouldMatchSnapshot = (title, component) => {
  it(title, done => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    renderWithHelmet(component).then(({ container }) => {
      const hasOneChild = container.children.length === 1;
      /*
       * if the container has more than one child then it's a component that uses a
       * fragment at the top level so we should not select the first child because it
       * wouldn't snapshot the whole component
       */
      expect(
        hasOneChild ? removeWrappingDiv(container) : container,
      ).toMatchSnapshot();

      done();
    });
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
