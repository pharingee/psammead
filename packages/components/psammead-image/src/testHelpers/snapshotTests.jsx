import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { custom, landscape, portrait, square } from './fixtureData';

const snapshotTests = (Component, additionalProps) => {
  shouldMatchSnapshot(
    'should render landscape image correctly',
    <Component
      alt={landscape.alt}
      attribution={landscape.attribution}
      sizes={landscape.sizes}
      src={landscape.src}
      height={landscape.height}
      width={landscape.width}
      {...additionalProps}
    />,
  );
  shouldMatchSnapshot(
    'should render portrait image correctly',
    <Component
      alt={portrait.alt}
      attribution={portrait.attribution}
      sizes={portrait.sizes}
      src={portrait.src}
      height={portrait.height}
      width={portrait.width}
      {...additionalProps}
    />,
  );
  shouldMatchSnapshot(
    'should render square image correctly',
    <Component
      alt={square.alt}
      attribution={square.attribution}
      sizes={square.sizes}
      src={square.src}
      height={square.height}
      width={square.width}
      {...additionalProps}
    />,
  );
  shouldMatchSnapshot(
    'should render image with custom dimensions correctly',
    <Component
      alt={custom.alt}
      attribution={custom.attribution}
      sizes={custom.sizes}
      src={custom.src}
      height={custom.height}
      width={custom.width}
      {...additionalProps}
    />,
  );
  shouldMatchSnapshot(
    'should render image with srcset correctly',
    <Component
      alt={landscape.alt}
      attribution={landscape.attribution}
      sizes={landscape.sizes}
      src={landscape.src}
      srcset={landscape.srcset}
      height={landscape.height}
      width={landscape.width}
      {...additionalProps}
    />,
  );
};

export default snapshotTests;
