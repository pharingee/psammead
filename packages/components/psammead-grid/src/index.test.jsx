import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import Grid from '.';
import { ExampleParagraph, ExampleFigure } from './testHelpers';

describe('Grid component', () => {
  shouldMatchSnapshot(
    'should render Grid with Grid items',
    <Grid
      columns={{ group3: 6, group4: 8, group5: 20 }}
      startOffset={{ group3: 1, group4: 1, group5: 5 }}
      enableGelMargins
      enableGelGutters
    >
      <Grid item columns={{ group3: 6, group4: 6, group5: 12 }}>
        <ExampleParagraph identifier="1" />
      </Grid>
      <Grid columns={{ group3: 6, group4: 6, group5: 12 }}>
        <Grid item columns={{ group3: 6, group4: 6, group5: 12 }}>
          <ExampleParagraph identifier="Landscape image " />
        </Grid>
        <Grid item columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}>
          <ExampleParagraph identifier="Landscape image's caption " />
        </Grid>
      </Grid>
      <Grid item columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}>
        <ExampleParagraph identifier="Paragraph " />
      </Grid>
      {['2', '3', '4', '5', '6', '7', '8', '9', '10'].map(num => (
        <Grid
          item
          columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}
          key={`${num}item`}
        >
          <ExampleParagraph identifier={num} />
        </Grid>
      ))}
    </Grid>,
  );
  shouldMatchSnapshot(
    'should render Grid with Grid items including nested non-Grid Figure element',
    <Grid columns={{ group3: 6, group4: 8, group5: 20 }} enableGelMargins>
      <Grid
        item
        columns={{ group3: 6, group4: 6, group5: 12 }}
        startOffset={{ group3: 1, group4: 1, group5: 5 }}
      >
        <ExampleParagraph identifier="1" />
      </Grid>
      <ExampleFigure>
        <Grid
          columns={{ group3: 6, group4: 6, group5: 12 }}
          startOffset={{ group3: 1, group4: 1, group5: 5 }}
        >
          <Grid
            item
            columns={{ group3: 6, group4: 6, group5: 12 }}
            startOffset={{ group3: 1, group4: 1, group5: 5 }}
          >
            <ExampleParagraph identifier="Landscape image " />
          </Grid>
          <Grid
            item
            columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}
            startOffset={{ group2: 1, group3: 1, group4: 1, group5: 5 }}
          >
            <ExampleParagraph identifier="Landscape image's caption " />
          </Grid>
        </Grid>
      </ExampleFigure>
      <Grid
        item
        columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}
        startOffset={{ group2: 1, group3: 1, group4: 1, group5: 5 }}
      >
        <ExampleParagraph identifier="Paragraph " />
      </Grid>
      {['2', '3', '4', '5', '6', '7', '8', '9', '10'].map(num => (
        <Grid
          item
          columns={{ group2: 6, group3: 5, group4: 5, group5: 10 }}
          startOffset={{ group2: 1, group3: 1, group4: 1, group5: 5 }}
          key={`${num}item`}
        >
          <ExampleParagraph identifier={num} />
        </Grid>
      ))}
    </Grid>,
  );
});
