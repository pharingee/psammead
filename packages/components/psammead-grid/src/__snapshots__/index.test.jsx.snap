// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component should render Grid with Grid items 1`] = `
.c0 {
  position: relative;
}

.c3 {
  position: relative;
}

.c2 {
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  color: #3F3F42;
  padding-bottom: 1.5rem;
  margin: 0;
}

@supports (display:grid) {
  .c0 {
    display: grid;
    position: initial;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c1 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c1 {
    display: inline-block;
    width: 75%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c1 {
    display: inline-block;
    width: 60%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c1 {
    display: block;
  }
}

@supports (display:grid) {
  .c3 {
    display: grid;
    position: initial;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c4 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c4 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c4 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c4 {
    display: block;
  }
}

@media (min-width:25rem) and (max-width:37.4375rem) {
  .c5 {
    display: inline-block;
    width: NaN%;
    vertical-align: top;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c5 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c5 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c5 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c5 {
    display: block;
  }
}

@media (min-width:25rem) and (max-width:37.4375rem) {
  .c6 {
    display: inline-block;
    width: NaN%;
    vertical-align: top;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c6 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c6 {
    display: inline-block;
    width: 62.5%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c6 {
    display: inline-block;
    width: 50%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c6 {
    display: block;
  }
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

@media (min-width:37.5rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p
      class="c2"
    >
      1
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c3"
  >
    <div
      class="c4"
    >
      <p
        class="c2"
      >
        Landscape image 
        This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
      </p>
    </div>
    <div
      class="c5"
    >
      <p
        class="c2"
      >
        Landscape image's caption 
        This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
      </p>
    </div>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      Paragraph 
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      2
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      3
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      4
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      5
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      6
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      7
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      8
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      9
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c6"
  >
    <p
      class="c2"
    >
      10
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
</div>
`;

exports[`Grid component should render Grid with Grid items including nested non-Grid Figure element 1`] = `
.c0 {
  position: relative;
}

.c4 {
  position: relative;
}

.c2 {
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  color: #3F3F42;
  padding-bottom: 1.5rem;
  margin: 0;
}

.c3 {
  margin: 0;
  padding: 0;
}

@supports (display:grid) {
  .c0 {
    display: grid;
    position: initial;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c1 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c1 {
    display: inline-block;
    width: 75%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c1 {
    display: inline-block;
    width: 60%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c1 {
    display: block;
  }
}

@supports (display:grid) {
  .c4 {
    display: grid;
    position: initial;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c5 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c5 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c5 {
    display: inline-block;
    width: 100%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c5 {
    display: block;
  }
}

@media (min-width:25rem) and (max-width:37.4375rem) {
  .c6 {
    display: inline-block;
    width: NaN%;
    vertical-align: top;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c6 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c6 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c6 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c6 {
    display: block;
  }
}

@media (min-width:25rem) and (max-width:37.4375rem) {
  .c7 {
    display: inline-block;
    width: NaN%;
    vertical-align: top;
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c7 {
    display: inline-block;
    width: 83.33333333333333%;
    vertical-align: top;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c7 {
    display: inline-block;
    width: 62.5%;
    vertical-align: top;
  }
}

@media (min-width:80rem) {
  .c7 {
    display: inline-block;
    width: 50%;
    vertical-align: top;
  }
}

@supports (display:grid) {
  .c7 {
    display: block;
  }
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

@media (min-width:37.5rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p
      class="c2"
    >
      1
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <figure
    class="c3"
  >
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <p
          class="c2"
        >
          Landscape image 
          This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
        </p>
      </div>
      <div
        class="c6"
      >
        <p
          class="c2"
        >
          Landscape image's caption 
          This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
        </p>
      </div>
    </div>
  </figure>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      Paragraph 
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      2
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      3
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      4
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      5
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      6
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      7
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      8
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      9
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
  <div
    class="c7"
  >
    <p
      class="c2"
    >
      10
      This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines. This is a long paragraph that will wrap for several lines.
    </p>
  </div>
</div>
`;
